mullvad-relay-selector/src/relay_selector/mod.rs

```diff
@@ -0,0 +1,853 @@
+//! The implementation of the relay selector.
+
+mod detailer;
+mod helpers;
+mod matcher;
+mod parsed_relays;
+pub mod query;
+
+use chrono::{DateTime, Local};
+use itertools::Itertools;
+use once_cell::sync::Lazy;
+use std::{
+    path::Path,
+    sync::{Arc, Mutex},
+    time::SystemTime,
+};
+
+use matcher::{BridgeMatcher, RelayMatcher, WireguardMatcher};
+use mullvad_types::{
+    constraints::Constraint,
+    custom_list::CustomListsSettings,
+    endpoint::MullvadWireguardEndpoint,
+    location::{Coordinates, Location},
+    relay_constraints::{
+        BridgeSettings, BridgeState, InternalBridgeConstraints, ObfuscationSettings,
+        OpenVpnConstraints, RelayOverride, RelaySettings, ResolvedBridgeSettings,
+        SelectedObfuscation, WireguardConstraints,
+    },
+    relay_list::{Relay, RelayList},
+    settings::Settings,
+    CustomTunnelEndpoint,
+};
+use talpid_types::{
+    net::{
+        obfuscation::ObfuscatorConfig, proxy::CustomProxy, Endpoint, TransportProtocol, TunnelType,
+    },
+    ErrorExt,
+};
+
+use crate::error::{EndpointError, Error};
+
+use self::{
+    detailer::{OpenVpnDetailer, WireguardDetailer},
+    matcher::AnyTunnelMatcher,
+    parsed_relays::ParsedRelays,
+    query::{BridgeQuery, Intersection, OpenVpnRelayQuery, RelayQuery, WireguardRelayQuery},
+};
+
+/// [`RETRY_ORDER`] defines an ordered set of relay parameters which the relay selector should prioritize on
+/// successive connection attempts. Note that these will *never* override user preferences.
+/// See [the documentation on `RelayQuery`][RelayQuery] for further details.
+///
+/// This list should be kept in sync with the expected behavior defined in `docs/relay-selector.ms`
+pub static RETRY_ORDER: Lazy<Vec<RelayQuery>> = Lazy::new(|| {
+    use query::builder::{IpVersion, RelayQueryBuilder};
+    vec![
+        // 0
```

Comment:
Although I am a proponent of zero-indexing, it is slightly confusing when comparing this list to the one in `relay-selector.md`. I think you should change it to match.